orangecontrib/text/widgets/owimporttexts.py

```diff
@@ -0,0 +1,642 @@
+            startdir=startdir,
+            future=None,
+            watcher=None,
+            cancelled=False,
+            cancel=None,
+        )
+
+        def cancel():
+            # Cancel the task and disconnect
+            if taskstate.future.cancel():
+                pass
+            else:
+                taskstate.task.cancelled = True
+                taskstate.cancelled = True
+                try:
+                    taskstate.future.result(timeout=10)
+                except UserInterruptError:
+                    pass
+                except TimeoutError:
+                    log.info("The task did not stop in in a timely manner")
+            taskstate.watcher.finished.disconnect(self.__onRunFinished)
+
+        taskstate.cancel = cancel
+
+        def run_text_scan_task_interupt():
+            try:
+                return task.run()
+            except UserInterruptError:
+                # Suppress interrupt errors, so they are not logged
+                return
+
+        taskstate.future = self.__executor.submit(run_text_scan_task_interupt)
+        taskstate.watcher = FutureWatcher(taskstate.future)
+        taskstate.watcher.finished.connect(self.__onRunFinished)
+
+    @Slot()
+    def __onRunFinished(self):
+        assert QThread.currentThread() is self.thread()
+        assert self.__state == State.Processing
+        assert self.__pendingTask is not None
+        assert self.sender() is self.__pendingTask.watcher
+        assert self.__pendingTask.future.done()
+        task = self.__pendingTask
+        self.__pendingTask = None
+
+        try:
+            text_data = task.future.result()
+        except Exception as err:
+            sys.excepthook(*sys.exc_info())
+            state = State.Error
+            text_data = []
+            self.error(traceback.format_exc())
+        else:
+            state = State.Done
+            self.error()
+
+        self._text_data = text_data
+        self._text_categories = TextScan.get_categories(text_data)
+
+        self.__setRuntimeState(state)
+        self.commit()
+
+    def cancel(self):
+        """
+        Cancel current pending task (if any).
+        """
+        if self.__state == State.Processing:
+            assert self.__pendingTask is not None
+            self.__pendingTask.cancel()
+            self.__pendingTask = None
+            self.__setRuntimeState(State.Cancelled)
+
+    @Slot(object)
+    def __onReportProgress(self, arg):
+        # report on scan progress from a worker thread
+        # arg must be a namespace(count: int, lastpath: str)
+        assert QThread.currentThread() is self.thread()
+        if self.__state == State.Processing:
+            self.pathlabel.setText(prettifypath(arg.lastpath))
+
+    def commit(self):
+        """
+        Create and commit a Corpus from the collected text meta data.
+        """
+        corpus = None
+        if self._text_data:
+            names = ["name", "path", "content"]
```

Comment:
Scripting is now supported. Now we need to choose a proper name.

- `TextScan` (current)
- `RetrieveText`
- `GetText`